File: src/commands/router.ts
import type { CommandHandler } from "./types.js";

type SyncCommandHandler = (args: string, channelId: string) => string;
type AsyncCommandHandler = (args: string, channelId: string) => Promise<string>;
type CommandHandlerFn = SyncCommandHandler | AsyncCommandHandler;

export class CommandRouter {
  private handlers = new Map<string, CommandHandlerFn>();

  register(prefix: string, handler: CommandHandlerFn): void {
    this.handlers.set(prefix, handler);
  }

  /**
   * Attempt to handle a Discord message as a command.
   * Returns the response string if handled, or null if not a command.
   */
  async handle(prompt: string, channelId: string): Promise<string | null> {
    for (const [prefix, handler] of this.handlers) {
      if (prefix === prompt || prompt.startsWith(prefix + " ")) {
        const args = prompt === prefix ? "" : prompt.slice(prefix.length + 1);
        return await handler(args, channelId);
      }
      // Handle commands like "!clear" that match exactly
      if (prompt === prefix) {
        return await handler("", channelId);
      }
    }
    return null;
  }
}

File: src/commands/router.py
from typing import Callable, Awaitable

CommandHandlerFn = Callable[[str, str, str], Awaitable[str | None]]


class CommandRouter:
    def __init__(self) -> None:
        self._handlers: dict[str, CommandHandlerFn] = {}

    def register(self, prefix: str, handler: CommandHandlerFn) -> None:
        self._handlers[prefix] = handler

    def is_command(self, text: str) -> bool:
        return any(text.startswith(p) for p in self._handlers)

    async def dispatch(
        self, text: str, channel_id: str, user_id: str
    ) -> str | None:
        for prefix, handler in self._handlers.items():
            if text.startswith(prefix):
                args = text[len(prefix):].strip()
                return await handler(channel_id, user_id, args)
        return None
